### PR TITLE
Extract shareable classes, improve interfaces

### PR DESCRIPTION
**Pull request by pbrisbin**
- Extract EnginesRunner
- Remove unused EngineProcess autoload
- Engine accepts non-string json configuration
- Sharable PathPatterns class
- Support dev_mode directly in EngineRegistry

/cc @codeclimate/review

### CODE REVIEW

**Comment by brynary** [comment]
@pbrisbin can we fixup the codeclimate issues detected (with the possible exception of the "long method" one... that's more subjective)?

**Comment by pbrisbin** [comment]
Always.

File: lib/cc/cli/analyze.rb [hunk 1]
```diff
@@ -1,11 +1,9 @@
-require "securerandom"
-
 module CC
   module CLI
     class Analyze < Command
       include CC::Analyzer
 
-      def initialize(args = [])
+      def initialize(_args = [])
```
Comment (GordonDiggs) [inline lib/cc/cli/analyze.rb]: What's the value of this change?

File: lib/cc/analyzer.rb [hunk 2]
```diff
@@ -5,23 +5,32 @@ module Analyzer
...
+    end
+
+    class DummyLogger
+      def method_missing(*)
         yield if block_given?
       end
     end
 
     cattr_accessor :statsd
+    cattr_accessor :logger
```
Comment (GordonDiggs) [inline lib/cc/analyzer.rb]: Any reason not to declare these on the same line?

File: lib/cc/cli/analyze.rb (hunk 1 above)
Comment (pbrisbin) [inline lib/cc/cli/analyze.rb]: It was a drive-by fix: `args` is unused, if this were written today, it would be a rubocop issue.

File: lib/cc/analyzer.rb (hunk 2 above)
Comment (pbrisbin) [inline lib/cc/analyzer.rb]: No reason, didn't think about it.

File: lib/cc/analyzer/engines_runner.rb [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+          engine_config = config.merge!(exclude_paths: exclude_paths)
+
+          Engine.new(name, metadata, @source_dir, engine_config, label)
+        end
+      end
+
+      def enabled_engines
+        {}.tap do |ret|
+          @config.engines.each do |name, config|
+            if config["enabled"]
```
Comment (pbrisbin) [inline lib/cc/analyzer/engines_runner.rb]: I think this should be `config.enabled?`, it's a true `CC::Yaml` object now.

File: lib/cc/analyzer/path_patterns.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+      private
+
+      def expand
+        results = Dir.chdir(@root) do
+          @patterns.map do |pattern|
+            Dir.glob(pattern)
+          end
+        end
+
+        results.flatten.sort.uniq
```
Comment (pbrisbin) [inline lib/cc/analyzer/path_patterns.rb]: Could remove `flatten` by using `flat_map` above.

Comment (brynary) [inline lib/cc/analyzer/path_patterns.rb]: Someone should add a rubocop check for flat_map :)



—Bryan Helmkamp, Founder, Code Climate

bryan@brynary.com / 646-379-1810 / @brynary




Typed with thumbs.

File: lib/cc/analyzer/engine_registry.rb [hunk 5]
```diff
@@ -3,18 +3,29 @@
...
-      def initialize
+      def initialize(dev_mode = false)
         @path = File.expand_path("../../../../config/engines.yml", __FILE__)
         @config = YAML.safe_load_file(@path)
+        @dev_mode = dev_mode
       end
 
       def [](engine_name)
-        @config[engine_name]
+        if dev_mode?
```
Comment (GordonDiggs) [inline lib/cc/analyzer/engine_registry.rb]: Just putting this out there, but I don't think it warrants a change just yet. I'm nervous about this being included in builder because of the potential for arbitrary docker containers to be run. WDYT of overriding this method within builder when we include this to always be false?

**Comment by GordonDiggs** [comment]
LGTM
